NXP backend: Add NeutronQuantizer

skywall opened the pull request:
### Summary
Implementation on NeutronQuantizer that quantizes model in a way optimized for Neutron NPU.


cc @digantdesai @JakeStevens @robert-kalmar

skywall committed:
NXP backend: Add NeutronQuantizer

skywall commented:
@pytorchbot label module:nxp

skywall commented:
@pytorchbot label "module: nxp"

skywall commented:
@pytorchbot label "release notes: nxp"

digantdesai commented on backends/nxp/quantizer/neutron_quantizer.py:
```diff
@@ -0,0 +1,244 @@
+    ) -> torch.fx.GraphModule:
+        return model
+
+    def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
```
Do we need these?

digantdesai commented on backends/nxp/tests/test_quantizer.py:
```diff
@@ -0,0 +1,204 @@
+    assert len(nodes) == 11
+    assert nodes[7].name == "conv2d"
+    # [0]: Input, [1] : weights, [2]: bias
+    assert _get_target_name(nodes[7].args[0]) == 'torch.ops.quantized_decomposed.dequantize_per_tensor.default'
```
Other option would be to use `FileCheck` utils, not a strong preference.

skywall replied:
Thanks for the tip. We plan to utilize `FileCheck` through some `ArmTester`-like interface, but we're not there yet. `FileCheck` in general doesn't provide an API to check order and specific index of an operators, what is at this stage important for us, because we're (not ideally) using models with multiple operators in tests.

digantdesai replied:
`ArmTester` derived from `XNNPACKTester`, alas same story as Quantizer. But since you don't have it implemented yet, can you please refactor `XNNPACKTester`, abstract parts of it as `BackendTester` and move of it to `backends/test` dir and make `XNNPACKTester`, `ArmTester` (already inherits from `XNNPACKTester`), and `NeutronTester` (new!) derive from that common tester? 🙏

skywall replied:
Sounds good to me. Should I open Github issue or we can track this internally?

JakeStevens replied:
Please do open a Github issue to track

skywall replied:
https://github.com/pytorch/executorch/issues/10100

digantdesai commented on backends/nxp/tests/models.py:
```diff
@@ -0,0 +1,221 @@
+# Copyright 2024 NXP
```
Other code organization can be `conv2d.py` which would define the model and implements relevant tests around it. This is also OK, just a suggestions, not a strong preference.

skywall replied:
Models in this file will be used in upcoming PRs and are reused by tests of different components. Tests are designed more as a "feature-based" - tests of single feature on multiple models, instead of "model-based" - various tests of single model. We'll split split the models into separate files in future PRs if you're fine with that.

digantdesai commented on backends/nxp/tests/test_quantizer.py:
```diff
@@ -0,0 +1,204 @@
+    assert _get_target_name(nodes[7].args[1]) == 'torch.ops.quantized_decomposed.dequantize_per_tensor.default'
+    assert _get_target_name(nodes[7].args[2]) == 'torch.ops.quantized_decomposed.dequantize_per_tensor.default'
+    assert _get_target_name(nodes[8]) == 'torch.ops.quantized_decomposed.quantize_per_tensor.default'
+    assert nodes[8].args[0].name == "conv2d"
```
can we also run it? Just to make sure the prepare/convert worked ok when consuming annotations, and the graph is still "functional"

skywall replied:
Thanks for the tip. Do you consider running `m(*example_input)` as enough in this case? As I have mentioned in previous comment, we plan to introduce some kind of `Tester` but we don't have this ready yet.

digantdesai reviewed (commented):
Thanks for the contribution. 

Looks good to me at a high level. Left a couple of comments, lets resolve this. We are quite close to accept/merging this.

JakeStevens commented:
@skywall  waiting on any updates

skywall commented on backends/nxp/quantizer/neutron_quantizer.py:
```diff
@@ -0,0 +1,244 @@
+        prev_node = fused_partition[0].input_nodes[0]
+
+        # In the case of a node with shared quantization spec has no previous node, return None to not quantize the node
+        if not hasattr(prev_node, "meta") or "quantization_annotation" not in prev_node.meta:
```
The comment in the code was a bit misleading and first part of the check unnecessary (`not hasattr(prev_node, "meta")`). Prev node will be always there. Updated.

skywall commented on backends/nxp/quantizer/neutron_quantizer.py:
```diff
@@ -0,0 +1,244 @@
+    Conv2dPattern,
+    LinearPattern,
+)
+from executorch.backends.cadence.aot.quantizer.quantizer import CadenceAtenQuantizer
```
Thank you for thorough description and possible approaches to improve this. I have decided to go with (B) - remove our dependency on Cadence quantizer at a cost of copying some lines of code. I expect we'll update quantizer a bit in a future what can give us an idea how to design some universal one (approach A).

mcremon-meta replied:
@skywall I think we can fairly easily refactor this in a nice way so that all base classes are shared and both quantizers derive from them in a way that minimize any major dependencies/risks. We very rarely update base classes anymore, and if that were to happen, we can do a good reviewing job here on GH and I don't anticipate issues. This PR is completely fine, but if there is consensus we will try to find eng time to merge those properly (cc @digantdesai @JakeStevens)

digantdesai replied:
Good discussion, thanks @JakeStevens, @mcremon-meta, @skywall.

I don't have a strong preference here. We can go as is and throw this down the line, or copy it now, or lastly, refactor Cadence (and potentially others) and create a better abstraction. 

We already have a [few](https://github.com/search?q=repo%3Apytorch%2Fexecutorch+%2Fclass+.*Quantizer%2F&type=code) quantizers in ET, so if I were you I would just copy it for now and create an issue to refactor this later. 

As others said it shouldn't be too difficult and potentially be useful to new delegates in the future. That said, I am a bit hesitant to put that as a blocker on this PR because it can potentially impact not just Cadence but other delegates too if you want to scrub a bit deeper.

skywall commented on backends/nxp/quantizer/neutron_quantizer.py:
```diff
@@ -0,0 +1,244 @@
+    is_dynamic=False,
+    observer_or_fake_quant_ctr=MinMaxObserver,
+)
+# Bias Quantization Specification is as follows:
+#   dtype = torch.int32
+#   quant_min, quant_max - full int32 range
+#   qcheme = torch.per_channel_symetric (for Conv), torch.per_tensor_symetric for Addmmn ==> i.e. zero_point = 0
+#   scale = input_scale * weight_scale
```
I'm sorry but I don't see the difference 🙈

digantdesai replied:
Remove the comments :p

skywall replied:
![](https://c.tenor.com/HdCNuyIa2c4AAAAC/tenor.gif)

skywall commented:
@digantdesai @JakeStevens Thank you for quick review and my apologies for late response. I hope I have addressed all your comments and I'm looking forward your feedback.

JakeStevens commented:
Looks like you need to run the linter and resubmit @skywall 

FWIW, I think this is a good state and we can work on reviewing the next chunk in parallel if there is more to submit

JakeStevens commented on backends/nxp/quantizer/neutron_quantizer.py:
```diff
@@ -0,0 +1,206 @@
+    QuantizationConfig,
+    QuantizationSpec,
+)
+from executorch.src.executorch.backends.nxp.quantizer.patterns import AddmmPattern
```
duplicate include still hitting linter

skywall replied:
Sorry, local lintrunner out of sync :( Fixed.

robert-kalmar commented:
> Looks like you need to run the linter and resubmit @skywall
> 
> FWIW, I think this is a good state and we can work on reviewing the next chunk in parallel if there is more to submit

@JakeStevens I am already working on the next chunk putting the Neutron Backend (i.e. the Delegate and the Partitioner) on the PR. Slightly more work to get it in sync with main, as of lot of new files added.

digantdesai reviewed (approved):
LGTM, thanks for the quick turnaround. Let's fix the CI and we can merge it.
